Add DFTB+ as variant

fangohr opened the pull request:
Addresses: https://github.com/fangohr/octopus-in-spack/issues/98

without MPI (input from Franco)

fangohr committed:
Add DFTB+ as variant

without MPI (input from Franco)

iamashwin99 commented:
the upstream dftb+ dosent compile, should we not wait for that to be fixed before setting it here?

iamashwin99 commented on spack/package.py:
```diff
@@ -127,6 +128,7 @@ class Octopus(AutotoolsPackage, CudaPackage):
     depends_on("libyaml", when="+libyaml")
     depends_on("pnfft", when="+pnfft")
     depends_on("nlopt", when="+nlopt")
+    depends_on("dftbplus~mpi", when="+dftbplus")
```
we need it as a shared lib and we also need to enable api(dont quite remember why the api is needed but thats the way it is in easybuild at least)
```suggestion
    depends_on("dftbplus~mpi+api+sharedlibs~openmp", when="+dftbplus")
```

fangohr replied:
None of `api`, `sharedlibs`, `openmp` exist as variants. Is this as different package? Or a significant update?

fangohr committed:
Input from Ashwin

Co-authored-by: Ashwin Kumar Karnad <46030335+iamashwin99@users.noreply.github.com>

fangohr commented:
Thanks, Ashwin. Why do we choose `~openmp`?

fangohr committed:
None of the variants from the last commit seem to exist

spack info dftbplus:

...

Variants:
    Name [Default]             When    Allowed values    Description
    =======================    ====    ==============    ===================================================================================================

    arpack [off]               --      on, off           Use ARPACK for excited state DFTB functionality
    build_system [makefile]    --      makefile          Build systems supported by the package
    dftd3 [off]                --      on, off           Use DftD3 dispersion library (if you need this dispersion model)
    elsi [off]                 --      on, off           Use the ELSI library for large scale systems. Only has any effect if you build with '+mpi'
    gpu [off]                  --      on, off           Use the MAGMA library for GPU accelerated computation
    mpi [on]                   --      on, off           Build an MPI-paralelised version of the code.
    sockets [off]              --      on, off           Whether the socket library (external control) should be linked
    transport [off]            --      on, off           Whether transport via libNEGF should be included. Only affects parallel build. (serial version is
...

fangohr committed:
choose the +dftbplus variant to actually build it

(might get a failure, see Ashwin's earlier comment)

fangohr commented:
> the upstream dftb+ doesnt compile, should we not wait for that to be fixed before setting it here?

Ah - I wasn't aware of that. Is there a MR (upstream in spack) to fix dftb+? (With my latest commit we should see the build failing: that would be my other question: what is the problem with dftb+ building in spack.)

iamashwin99 commented:
> > the upstream dftb+ doesnt compile, should we not wait for that to be fixed before setting it here?
> 
> Ah - I wasn't aware of that. Is there a MR (upstream in spack) to fix dftb+? (With my latest commit we should see the build failing: that would be my other question: what is the problem with dftb+ building in spack.)

https://github.com/spack/spack/issues/23086
https://github.com/spack/spack/pull/39608

fangohr commented:
Okay, let's put this on hold until the DFTB+ changes are merged upstream (either https://github.com/spack/spack/issues/23086 would be good https://github.com/spack/spack/pull/39608). The one from Ashwin would probably be 'better' for us as these changes are already in our local MPSD-spack fork, and the buildbot uses them.

lang-m commented:
@fangohr Should we maybe just close this and add DFTB+ support once we have finished the transition to CMake?
